src/frontend/src/lib/components/ui/Button.svelte

```diff
@@ -0,0 +1,37 @@
+<script lang="ts">
+	type HierarchyOptions =
+		| { primary: true; secondary?: never }
+		| { secondary: true; primary?: never };
+
+	// eslint-disable-next-line @typescript-eslint/no-unused-vars
+	type $$Props = HierarchyOptions & {
+		type?: 'submit' | 'reset' | 'button';
+		disabled?: boolean;
+	};
+
+	export let type: 'submit' | 'reset' | 'button' = 'submit';
+	export let disabled = false;
+
+	// We guarantee that, if either primary or secondary is set, they mutually exclude each other
```

Comment:
Not familliar with this pattern. What happens if not, does it fails at build time or runtime?

Build time sounds good. If runtime, how's the error communicated to the user?

Reply:
For now it is just type-notation-safe, it doesn't fail at runtime or buildtime. At run time, we fallback on defining both props if none is provided, or on primary if both provided (and the user ignored the type-warning).

So the cases are, user passes in a non-type-safe way:

1. `primary` provided --> `primary` is used and `secondary` is the opposite
2. `secondary` provided --> `secondary` is used and `primary` is the opposite
3. both provided --> `primary` takes precedence and we follow rules 1
4. none provided --> `primary` and `secondary` are false, so it's a normal button

Reply:
It's a (really) nice pattern but, if it's "just" for type safety kind of feel like it's an over-complication.

```
export let color: "primary" | "secondary" = "primary";
```

(or another) would do the job as well no?

Reply:
ah yes, agreed! I did like that because I thought we wanted a very generic `Button` component, but if it's only `primary` or `secondary`, then yes, for sure! Shall I do your suggestion then?

Reply:
Yes I would say so.

PS: the pattern is really cool! Just thinking we can keep it simple at least for now.

Reply:
yes yes! agreed! changed!

and BTW, the pattern is in the history now, so all good if we need it at some point eheh

Reply:
Nice!